datafusion-examples/examples/advanced_udf/main.rs

```diff
@@ -0,0 +1,78 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! # Advanced UDF/UDAF/UDWF/Asynchronous UDF Examples
+//!
+//! This example demonstrates advanced user-defined functions in DataFusion.
+//!
+//! ## Usage
+//! ```bash
+//! cargo run --example advanced_udf -- [udf|udaf|udwf|async_udf]
+//! ```
+//!
+//! Each subcommand runs a corresponding example:
+//! - `udf` — user defined scalar function example
+//! - `udaf` — user defined aggregate function example
+//! - `udwf` — user defined window function example
+//! - `async_udf` — asynchronous user defined function example
+
+mod async_udf;
+mod udaf;
+mod udf;
+mod udwf;
+
+use std::str::FromStr;
+
+use datafusion::error::{DataFusionError, Result};
+
+enum ExampleKind {
+    Udf,
+    Udaf,
+    Udwf,
+    AsyncUdf,
+}
+
+impl FromStr for ExampleKind {
```

Comment:
I do wonder if there's a way to templatize/abstract this whole file so they don't drift in structure from the others 🤔